Dependency Handling section in Library Guidelines (#10487)
elixir-lang/elixir: lib/elixir/pages/library-guidelines.md

The `mix new` command also allows the `--sup` option to scaffold an application with a supervision tree out of the box. We talk about supervision trees later on when discussing one of the common anti-patterns when writing libraries.

## Dependency handling

> When the library is used as a dependency, its [lockfile](https://hexdocs.pm/mix/Mix.Project.html#module-configuration) (usually named `mix.lock`) is _ignored by the host project_. Running `mix deps.get` in the host project attempts to get the latest possible versions of your library’s dependencies, as by pins in your `mix.exs`. These versions might be greater than those stored in your `mix.lock` (and hence used in your tests / CI).

Reviewer: ```suggestion
When the library is used as a dependency, its [lockfile](https://hexdocs.pm/mix/Mix.Project.html#module-configuration) (usually named `mix.lock`) is _ignored by the host project_. Running `mix deps.get` in the host project attempts to get the latest possible versions of your library’s dependencies, as specified by the requirements in the `deps` section of your `mix.exs`. These versions might be greater than those stored in your `mix.lock` (and hence used in your tests / CI).
```

> On the other hand, contributors of your library, need a deterministic build, which implies the presense of `mix.lock` in the VCS.

Reviewer: ```suggestion
On the other hand, contributors of your library, need a deterministic build, which implies the presense of `mix.lock` in your Version Control System (VCS).
```

> The best practice of handling `mix.lock` file therefore would be to keep it in VCS, and run two different CI workflows: the usual deterministic one, and another one, that starts with `mix deps.unlock --all` and always compiles your library and runs tests against latest versions of dependencies. The latter one might be even run nightly or otherwise recurrently to stay notified about any possible issue in regard to dependencies updates.

Reviewer: ```suggestion
The best practice of handling `mix.lock` file therefore would be to keep it in VCS, and run two different Continuous Integration (CI) workflows: the usual deterministic one, and another one, that starts with `mix deps.unlock --all` and always compiles your library and runs tests against latest versions of dependencies. The latter one might be even run nightly or otherwise recurrently to stay notified about any possible issue in regard to dependencies updates.
```
